src/jitlayers.cpp

```diff
@@ -197,47 +197,84 @@ static jl_callptr_t _jl_compile_codeinst(
+bool jl_type_mappable_to_c(jl_value_t *ty);
+
+// declare a C-callable entry point; called during code loading from the toplevel
 extern "C" JL_DLLEXPORT
-void jl_extern_c(jl_function_t *f, jl_value_t *rt, jl_value_t *argt, char *name)
-{
+void jl_extern_c(jl_value_t *declrt, jl_tupletype_t *sigt)
+{
+    // validate arguments. try to do as many checks as possible here to avoid
+    // throwing errors later during codegen.
+    JL_TYPECHK(@ccallable, type, declrt);
+    if (!jl_is_tuple_type(sigt))
+        jl_type_error("@ccallable", (jl_value_t*)jl_anytuple_type_type, (jl_value_t*)sigt);
+    // check that f is a guaranteed singleton type
+    jl_datatype_t *ft = (jl_datatype_t*)jl_tparam0(sigt);
+    if (!jl_is_datatype(ft) || ft->instance == NULL)
+        jl_error("@ccallable: function object must be a singleton");
+
+    // compute / validate return type
+    jl_value_t *crt = declrt;
+    if (jl_is_abstract_ref_type(declrt)) {
+        jl_value_t *declrt0 = jl_tparam0(declrt);
+        if (jl_is_typevar(declrt0))
+            jl_error("@ccallable: return type Ref should have an element type, not Ref{<:T}");
+        if (declrt0 == (jl_value_t*)jl_any_type)
+            jl_error("@ccallable: return type Ref{Any} is invalid. Use Any or Ptr{Any} instead.");
+        crt = (jl_value_t*)jl_any_type;
+    }
     JL_LOCK(&codegen_lock);
-    jl_codegen_params_t params;
-    Function *llvmf = jl_cfunction_object(f, rt, (jl_tupletype_t*)argt, params);
-    jl_jit_globals(params.globals);
-    assert(params.workqueue.empty());
-    // force eager emission of the function (llvm 3.3 gets confused otherwise and tries to do recursive compilation)
-    jl_add_to_ee();
-    uint64_t Addr = getAddressForFunction(llvmf->getName());
-
-    if (imaging_mode)
-        llvmf = cast<Function>(shadow_output->getNamedValue(llvmf->getName()));
-
-    // make the alias to the shadow_module
-    GlobalAlias *GA =
-        GlobalAlias::create(llvmf->getType()->getElementType(), llvmf->getType()->getAddressSpace(),
-                            GlobalValue::ExternalLinkage, name, llvmf, shadow_output);
-
-    // make sure the alias name is valid for the current session
-    jl_ExecutionEngine->addGlobalMapping(GA, (void*)(uintptr_t)Addr);
+    if (!jl_type_mappable_to_c(crt))
+        jl_error("@ccallable: return type doesn't correspond to a C type");
     JL_UNLOCK(&codegen_lock);
+
+    // validate method signature
+    size_t i, nargs = jl_nparams(sigt);
+    for (i = 1; i < nargs; i++) {
+        jl_value_t *ati = jl_tparam(sigt, i);
+        if (jl_is_abstract_ref_type(ati)) {
+            ati = jl_tparam0(ati);
+            if (jl_is_typevar(ati))
+                jl_error("@ccallable: argument type Ref should have an element type, not Ref{<:T}");
+        }
+        if (jl_is_pointer(ati) && jl_is_typevar(jl_tparam0(ati)))
+            jl_error("@ccallable: argument type Ptr should have an element type, Ptr{<:T}");
+    }
```

Comment:
Ah, I see this design is just handling all of the `Ref` declared signatures incorrectly. It seems part also of why they can't appropriately be attached to a single Method.

Reply:
Yeah, I would be fine with not supporting `Ref` at all here --- no julia values, no by-reference, only simple types you'd write directly in C code. That would make a lot of this easier.